### PR TITLE
Quote arguments with special characters (e.g. "&") on Windows

### PR DESCRIPTION
**Pull request by malthe**
This fixes issue #35124. Note that it is quite unlikely that this issue would be mitigated or fixed in the Python standard library itself code due to the history of "subprocess.list2cmdline" (see https://bugs.python.org/issue8972 which was created in 2010).

### What changes were proposed in this pull request?

This adds support for a number of special characters (such as "&") appearing in Spark configuration values (common in cloud resource URIs).

### Why are the changes needed?

If one of the relevant characters appears in a configuration value, the Windows submit scripts do not work correctly, interpreting the character literally. For example, an ampersand marks the beginning of a new command (leading Windows to actually start multiple processes which would typically fail with a rather strange and unexpected error message).

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

This was tested manually, providing a configuration value with a "&" character in it.

### CODE REVIEW

**Comment by malthe** [comment]
I have added support for the remaining set of characters: "&<>()|,;^".

**Comment by holdenk** [comment]
Jenkins ok to test

File: python/pyspark/java_gateway.py [hunk 1]
```diff
@@ -97,8 +99,18 @@ def preexec_func():
...
                 proc = Popen(command, **popen_kwargs)
             else:
+                # If an argument contains a character that needs escaping (e.g. an 
+                # ampersand) we need to apply a double quoting mechanism to correctly 
+                # pass the value through the submit shell scripts.
+                def quote(s):
+                    r, count = subn(r"([&<>()|,;^])", r"^^^\1", s)
+                    if count:
+                        return '"""' + r + '"""'
+                    return s
```
Comment (holdenk) [inline python/pyspark/java_gateway.py:109]: Can we put this in a utils function and or is there a library that does this escaping? I imagine there's a bunch of random values and I'd rather us get them all.

**Review by holdenk — COMMENTED** [review]
Thanks for working on this.  Also is there a link to explaining the popen on windows issue? I'm not a windows user so I'd like some more understanding.

File: python/pyspark/java_gateway.py (hunk 1 above)
Comment (malthe) [reply python/pyspark/java_gateway.py:109]: I have looked and not found any ...

**Comment by malthe** [comment]
@holdenk I'm not really a Windows user either so I had to piece together a solution based on snippets and assorted help pages from Microsoft – really I just ended up experimenting to get it right.

**Pull request closed without merging**